source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/Proxies/ServiceProxy.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+
+namespace Octopus.Tentacle.Tests.Integration.Util.Builders.Decorators.Proxies
+{
+    public abstract class ServiceProxy : DispatchProxyAsync
+    {
+        protected object? TargetService { get; private set; }
+
+        protected void SetTargetService(object service)
+        {
+            TargetService = service;
+        }
+
+        public override object Invoke(MethodInfo method, object[] args)
+        {
+            EnsureTargetServiceNotNull();
+            try
+            {
+                OnStartingInvocation(method);
+
+                return method.Invoke(TargetService, args)!;
+            }
+            catch (TargetInvocationException e)
+            {
+                OnInvocationException(method, e.InnerException!);
+
+                //we need to unwrap the TargetInvocationException
+                ExceptionDispatchInfo.Capture(e.InnerException).Throw();
+
+                //this will never be hit because the line above will throw an exception
+                return default;
+            }
+            catch (Exception e)
+            {
+                OnInvocationException(method, e);
+                throw;
+            }
+            finally
+            {
+                OnCompletingInvocation(method);
+            }
+        }
+
+        public override async Task InvokeAsync(MethodInfo method, object[] args)
+        {
+            EnsureTargetServiceNotNull();
+            try
+            {
+                await OnStartingInvocationAsync(method).ConfigureAwait(false);
+
+                var task = (Task)method.Invoke(TargetService, args);
+
+                await task!.ConfigureAwait(false);
+            }
+            catch (TargetInvocationException e)
+            {
+                OnInvocationException(method, e.InnerException!);
+
+                //we need to unwrap the TargetInvocationException
+                ExceptionDispatchInfo.Capture(e.InnerException).Throw();
+            }
+            catch (Exception e)
+            {
+                OnInvocationException(method, e);
+                throw;
+            }
```

Comment:
Does this need a finally here? This looks like one is intended

Reply:
ooops, yep. thanks!